### PR TITLE
[SDK] Feature: Enclave wallets

### PR DESCRIPTION
**Pull request by gregfromstl**
## Problem solved

Short description of the bug fixed or feature added

---

## PR-Codex overview
This PR focuses on enhancing enclave wallet functionality. 

### Detailed summary
- Added `IWebWallet` interface with methods
- Modified wallet setup and authentication processes
- Implemented actions for enclave wallet management

> The following files were skipped due to too many changes: `packages/thirdweb/src/wallets/in-app/web/lib/auth/iframe-auth.ts`, `packages/thirdweb/src/wallets/in-app/web/lib/enclave-wallet.ts`, `packages/thirdweb/src/wallets/in-app/web/lib/web-connector.ts`

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

File: packages/thirdweb/src/wallets/in-app/web/lib/auth/iframe-auth.ts [hunk 1]
```diff
@@ -102,15 +106,55 @@ export class Auth {
...
+        params: {
+          storedToken: authToken.storedToken,
+          recoveryCode,
+        },
+      });
+      return this.postLogin(result);
+    }
+  }
+
+  async generateWallet(authResult: AuthStoredTokenWithCookieReturnType) {
```
Comment (joaquim-verges) [inline packages/thirdweb/src/wallets/in-app/web/lib/auth/iframe-auth.ts:132]: extract this out into /lib/actions

File: packages/thirdweb/src/wallets/in-app/web/lib/auth/iframe-auth.ts [hunk 2]
```diff
@@ -102,15 +106,55 @@ export class Auth {
...
+    const clientFetch = getClientFetch(this.client, this.ecosystem);
+    const response = await clientFetch(
+      `${getThirdwebBaseUrl("inAppWallet")}/api/v1/enclave-wallet/generate`,
+      {
+        method: "POST",
+        headers: {
+          "Content-Type": "application/json",
+          "x-thirdweb-client-id": this.client.clientId,
```
Comment (joaquim-verges) [inline packages/thirdweb/src/wallets/in-app/web/lib/auth/iframe-auth.ts:140]: we should not need this. cc @ElasticBottle

File: packages/thirdweb/src/wallets/in-app/web/lib/auth/oauth.ts [hunk 3]
```diff
@@ -100,6 +100,7 @@ export const loginWithOauth = async (options: {
               win,
               closeOpenedWindow: options.closeOpenedWindow,
             });
+            console.log("oauthSuccessResult", event.data.authResult);
```
Comment (joaquim-verges) [inline packages/thirdweb/src/wallets/in-app/web/lib/auth/oauth.ts:103]: cleanup

File: packages/thirdweb/src/wallets/in-app/web/lib/enclave-wallet.ts [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+    const client = this.client;
+    const ecosystem = this.ecosystem;
+
+    const _signTransaction = async (tx: SendTransactionOption) => {
+      const rpcRequest = getRpcClient({
+        client,
+        chain: getCachedChain(tx.chainId),
+      });
+      // biome-ignore lint/suspicious/noExplicitAny: ethers tx transformation
+      const transaction: Record<string, any> = {
```
Comment (joaquim-verges) [inline packages/thirdweb/src/wallets/in-app/web/lib/enclave-wallet.ts]: `Record<string, Hex | number | undefined>` ?

File: packages/thirdweb/src/wallets/in-app/web/lib/web-connector.ts [hunk 5]
```diff
@@ -111,6 +130,37 @@ export class InAppWebConnector implements InAppConnector {
...
   }
 
+  async initializeWallet(authToken?: string) {
+    const storedAuthToken = await getAuthToken(this.client, this.ecosystem);
+    if (!authToken && storedAuthToken === null) {
+        throw new Error("No auth token provided and no stored auth token found to initialize the wallet");
+    }
+
+    const user = await getEnclaveUserStatus({ authToken: authToken || storedAuthToken as string, client: this.client, ecosystem: this.ecosystem });
+    if (user.wallets[0].type === "enclave") {
```
Comment (joaquim-verges) [inline packages/thirdweb/src/wallets/in-app/web/lib/web-connector.ts:140]: user can be undefined? / wallets can be empty? make sure to guard against that

File: packages/thirdweb/src/wallets/in-app/web/lib/web-connector.ts [hunk 6]
```diff
@@ -111,6 +130,37 @@ export class InAppWebConnector implements InAppConnector {
     });
   }
 
+  async initializeWallet(authToken?: string) {
```
Comment (joaquim-verges) [inline packages/thirdweb/src/wallets/in-app/web/lib/web-connector.ts:133]: can you re-use this function in the auth callback

File: packages/thirdweb/src/wallets/in-app/web/utils/get-enclave-user-status.ts [hunk 7]
```diff
@@ -0,0 +1,29 @@
+import type { ThirdwebClient } from "../../../../client/client.js";
+import { getThirdwebBaseUrl } from "../../../../utils/domains.js";
+import { getClientFetch } from "../../../../utils/fetch.js";
+import type { UserStatus } from "../lib/enclave-wallet.js";
+import type { Ecosystem } from "../types.js";
+
+export async function getEnclaveUserStatus(
```
Comment (joaquim-verges) [inline packages/thirdweb/src/wallets/in-app/web/utils/get-enclave-user-status.ts:7]: can you also move this to /lib/actions? its nice to have all the endpoint wrappers in one place

File: packages/thirdweb/src/wallets/in-app/web/lib/enclave-wallet.ts [hunk 8]
```diff
@@ -0,0 +1,289 @@
...
+   */
+  async getUserWalletStatus(): Promise<GetUser> {
+    const clientFetch = getClientFetch(this.client, this.ecosystem);
+    const token = await this.localStorage.getAuthCookie();
+    if (!token) {
+      return { status: UserWalletStatus.LOGGED_OUT };
+    }
+
+    const response = await clientFetch(
+      `${getThirdwebBaseUrl("inAppWallet")}/api/2024-05-05/accounts`,
```
Comment (joaquim-verges) [inline packages/thirdweb/src/wallets/in-app/web/lib/enclave-wallet.ts:98]: you already have a function for this in get-enclave-user-status.ts

File: packages/thirdweb/src/wallets/in-app/web/lib/web-connector.ts (hunk 5 above)
Comment (gregfromstl) [reply packages/thirdweb/src/wallets/in-app/web/lib/web-connector.ts:140]: The backend _should_ always send exactly one wallet but I added a check in `getEnclaveUserStatus` to throw if that doesn't happen for some reason

File: apps/playground-web/src/components/styled-connect-button.tsx [hunk 9]
```diff
@@ -17,6 +18,7 @@ export function StyledConnectButton(
       supportedNFTs={{
         "84532": ["0x638263e3eAa3917a53630e61B1fBa685308024fa"],
       }}
+      chain={sepolia}
```
Comment (joaquim-verges) [inline apps/playground-web/src/components/styled-connect-button.tsx:21]: remove this, fixed default chains in my PR, this actually forces sepolia which we dont want

File: apps/playground-web/src/lib/constants.ts [hunk 10]
```diff
@@ -31,6 +31,7 @@ export const WALLETS = [
       passkeyDomain: getDomain(),
     },
   }),
+  ecosystemWallet("ecosystem.new-age"),
```
Comment (joaquim-verges) [inline apps/playground-web/src/lib/constants.ts:34]: also handled dev/prod in my PR (though commented)

we should make a new section in playground for ecosystem i reckon

File: packages/thirdweb/src/wallets/in-app/web/lib/auth/iframe-auth.ts [hunk 11]
```diff
@@ -280,15 +320,11 @@ export class Auth {
    * @internal
    */
   async logout(): Promise<LogoutReturnType> {
-    const { success } = await this.AuthQuerier.call<LogoutReturnType>({
```
Comment (joaquim-verges) [inline packages/thirdweb/src/wallets/in-app/web/lib/auth/iframe-auth.ts:283]: we sure this is safe to remove?

**Review by joaquim-verges — APPROVED** [review]
looks good, but double checking the logout thing

File: packages/thirdweb/src/wallets/in-app/web/lib/auth/iframe-auth.ts (hunk 11 above)
Comment (gregfromstl) [reply packages/thirdweb/src/wallets/in-app/web/lib/auth/iframe-auth.ts:283]: Yeah tested it pretty thoroughly

**Pull request merged**